### PR TITLE
 Fix hashing of NULL values

### LINKED ISSUE #48623: Hash on array with NULL throws exception
```sql
CREATE TABLE test
(
    c Array(Nullable(Int64))
)
ENGINE = Memory;

INSERT INTO test VALUES (NULL) ([-1,NULL]);

SELECT xxHash32(c) FROM test
```

Produces
```
Received exception from server (version 23.4.1):
Code: 48. DB::Exception: Received from localhost:9001. DB::Exception: Method getDataAt is not supported for Nullable(Int64) in case if value is NULL: while executing 'FUNCTION xxHash32(a :: 0) -> xxHash32(a) UInt32 : 1'. (NOT_IMPLEMENTED)
```

This used to work until #40195.

See also #51541

### PR DESCRIPTION
**Pull request by aiven-sal**
Closes #48623

This is a re-submit of #48625 with minor changes to make it backwards-compatible (see https://github.com/ClickHouse/ClickHouse/pull/48625#issuecomment-1872943188)

`NULL`s are assumed to have a hash value of 42, settling the question of life, the universe and everything.

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
It is now possible to hash `NULL` values in arbitrary `Nullable` types, e.g., `SELECT xxHash32([toNullable(NULL)]);`

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit d1f07bc9d1449b8b7921ba0c126df78d393a5603 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58754/d1f07bc9d1449b8b7921ba0c126df78d393a5603/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58754/d1f07bc9d1449b8b7921ba0c126df78d393a5603/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by zhanglistar** [comment]
@aiven-sal Sorry for not much time to do this, feed free to go ahead, thanks for continue the work.

File: tests/queries/0_stateless/02677_hash_functions_on_null.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
+select xxHash32(NULL);
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02677_hash_functions_on_null.sql:1]: Cosmetic: could we please put SQL keywords in all-lowercase or all UPPERCASE but not a mix of both?

File: tests/queries/0_stateless/02677_hash_functions_on_null.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+select xxHash64(NULL);
+
+DROP TABLE IF EXISTS test_hash_on_null;
+CREATE TABLE test_hash_on_null (a Array(Nullable(Int64))) ENGINE = Memory;
+insert into test_hash_on_null values (NULL) ([NULL, NULL]);
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02677_hash_functions_on_null.sql:6]: The column type is `Array(Nullable(T))`. What is the actually stored value if `NULL` is inserted? (also wondering, how that is even possible?)

Comment (rschu1ze) [inline tests/queries/0_stateless/02677_hash_functions_on_null.sql:6]: Can we add a test case which mixes non-NULL and NULL values inside an Array?

**Comment by rschu1ze** [comment]
> This is a re-submit of https://github.com/ClickHouse/ClickHouse/pull/48625 with minor changes to make it backwards-compatible (see https://github.com/ClickHouse/ClickHouse/pull/48625#issuecomment-1872943188)

@aiven-sal For future reference and for my curiosity, which backward-compatibility issues were addressed exactly and how?

**Comment by rschu1ze** [comment]
@aiven-sal Would you like to merge from master and have a look at the comments? I'd be happy to merge this one.

**Comment by alexey-milovidov** [comment]
Looks abandoned.

**Comment by Algunenano** [comment]
Please reopen once the comments have been addressed

**Pull request closed without merging**